### ISSUE TITLE
Limit columns retrieved from database

### ISSUE THREAD

**Issue by r2baka** [ardalis/Specification#238]
I'm trying to control the SELECT clause to prevent the retrieval of a large column from the database as so:

```
    public class DocumentsSpecification : Specification<Document, Document>
    {
        public DocumentsSpecification()
        {
            Query.Select(a => new Document(){ 
                Id = a.Id,
                Title = a.Title,
                Size = a.Size, 
                MimeType = a.MimeType
            }).OrderByDescending(a => a.UpdatedAt);
        }
    }
```
However, the query still retrieves ALL the fields, including the large [Data] in the SQL Statement, as seen here:
```
SELECT [d].[id], [d].[actual_size], [d].[created_at], 
            [d].[data], [d].[description], 
            [d].[extension], [d].[mime_type], [d].[size], [d].[title], 
            [d].[updated_at], [d].[url], [d].[uuid], [d].[visible]
FROM [documents] AS [d]
ORDER BY [d].[updated_at] DESC
```
Is what I'm trying to do possible? Just to try to understand better, I changed it to 
#
`                Title = a.Title + a.Title + a.Title,
`
But the title wasn't concatenated, I just got the single title, so I'm clearly not understanding something going on here.

Perhaps this is not a use case covered by the Specification?

**Comment by devbased** [comment]
Cannot reproduce (efcore) :thinking:
```csharp
public class StoreNameWithIdOrderedByCityDescendingSpec : Specification<Store, Store>
{
  public StoreNameWithIdOrderedByCityDescendingSpec()
  {
    Query.Select(x => new Store { Id = x.Id, Name = x.Name })
      .OrderByDescending(x => x.City);
  }
}

var result = await storeRepository.ListAsync(new StoreNameWithIdOrderedByCityDescendingSpec());

result.Should().NotBeNull();
```
```sql
SELECT [s].[Id], [s].[Name]
FROM [Stores] AS [s]
ORDER BY [s].[City] DESC
```
Concatenated
```sql
SELECT [s].[Id], (COALESCE([s].[Name], N'') + COALESCE([s].[Name], N'')) + COALESCE([s].[Name], N'') AS [Name]
FROM [Stores] AS [s]
ORDER BY [s].[City] DESC
```

**Comment by r2baka** [comment]
Thank you devbased, I'm diving in more here...

Stepping through just to be sure
![image](https://user-images.githubusercontent.com/101055/152659839-5cc2b53f-f640-419b-aa8d-dc5260c0b269.png)

Then, breaking in the Specification, I've commented everything but Title

![image](https://user-images.githubusercontent.com/101055/152659878-f8872673-d25a-46df-854e-87b9fc100677.png)

![image](https://user-images.githubusercontent.com/101055/152659898-bd2dcdc8-dbce-4576-a264-11727ce713ea.png)

after breaking on ListAsync
![image](https://user-images.githubusercontent.com/101055/152659914-993a0ee2-46d9-4058-9795-fb1d1c09f413.png)

The debug output shows all columns being retrieved.
![image](https://user-images.githubusercontent.com/101055/152659998-37a186a9-0555-4c0e-928d-8c02d29e78dc.png)

Just to be sure, I add the list to watches and drill into a document, and the Data field is populated, which verifies the data is being pulled back from the database.

![image](https://user-images.githubusercontent.com/101055/152660176-86327b63-dd2c-49e8-9857-ea5844e156e5.png)

I'm  using Ardalis.Specification 6.0.1
![image](https://user-images.githubusercontent.com/101055/152660259-7ba7c95f-e101-4c29-aba2-2d96a6dc6f92.png)

Thinking maybe it was my EntityFramework version, I tried upgrading that, but it isn't compatible with net5.0, could that be it?

Error	NU1202	Package Microsoft.EntityFrameworkCore.SqlServer 6.0.1 is not compatible with net5.0

I tested on a local SQL Server database and also Azure SQL Database in the cloud, both return full results.

Can I check somewhere else? Perhaps another package version?

Thanks again!

**Comment by devbased** [comment]
I changed project to net5.0 and installed Microsoft.EntityFrameworkCore.SqlServer 5.0.13 with 6.0.1 Specifications, still not reproduced.
Check your model, maybe you've included ignored properties and efcore switched to client-side evaluation.

**Comment by r2baka** [comment]
I upgraded to SqlServer 5.0.13, still doesn't work. The field for Data is nullable. Nothing odd there. I just wrote a custom SQL statement to select '' as data and that accomplishes the goal.

Thank you for your time helping with it. This work around is acceptable.

**Comment by fiseni** [comment]
Hey @r2baka,

Are you using our built-in repository or do you have your implementation? There should be two overloads for `ListAsync`
```c#
public async Task<List<T>> ListAsync(ISpecification<T> specification, CancellationToken cancellationToken = default)
public async Task<List<TResult>> ListAsync<TResult>(ISpecification<T, TResult> specification, CancellationToken cancellationToken = default)
```

First, ensure that you have that (refer to our implementation [here](https://github.com/ardalis/Specification/blob/main/Specification.EntityFrameworkCore/src/Ardalis.Specification.EntityFrameworkCore/RepositoryBaseOfT.cs)). Furthermore, since you're using the entity itself as return type `Specification<Document, Document>`, perhaps the call is not inferred correctly and the wrong `ListAsync` method is being utilized. Try to be explicit in the call, and see how it goes
```c#
var result = await documentRepository.ListAsync<Document>(new DocumentsSpecification());
```

**Comment by r2baka** [comment]
TLDR: That was it! But it took some modifications to make it work. Here's what I did.

I checked the Interface and I only have the first implementation:
`
public async Task<List<T>> ListAsync(ISpecification<T> specification, CancellationToken cancellationToken = default);
`
I think because I started from EShopOnWeb and my EFRepository looks like this one. I guess it hasn't been updated to use the latest Specifications. Last commit was from November 2020.
https://github.com/dotnet-architecture/eShopOnWeb/blob/6041a1f183e05618db8a26bedd3418795c6a7b2c/src/Infrastructure/Data/EfRepository.cs

I tried adding the second method to the IAsyncRepository, which required I add it to EFRepository, so I copy/pasted the one from [RepositoryBaseOfT.cs](https://github.com/ardalis/Specification/blob/main/Specification.EntityFrameworkCore/src/Ardalis.Specification.EntityFrameworkCore/RepositoryBaseOfT.cs) which I don't have in my project. But I got an error, so I did this:
```
        public virtual async Task<IReadOnlyList<TResult>> ListAsync<TResult>(ISpecification<T, TResult> specification, CancellationToken cancellationToken = default)
        {
            var specificationResult = ApplySpecification(specification);
            return await specificationResult.ToListAsync(cancellationToken);
        }
```
A little more investigation and I found I needed a new ApplySpecification, so I added this one to my EfRepository:
```
        protected virtual IQueryable<TResult> ApplySpecification<TResult>(ISpecification<T, TResult> specification)
        {
            var evaluator = new SpecificationEvaluator();
            return evaluator.GetQuery(_dbContext.Set<T>().AsQueryable(), specification);
        }
```

And now this works as expected

```
            DocumentsSpecification spec = new();
            var x = await _itemRepository.ListAsync(spec);
            DocumentIndexViewModel d = new();
            d.DocumentItems = x.ToJson().FromJson<List<DocumentItemViewModel>>();
            return d;

```

That was a fairly easy fix and thought it would be easier than updating the entire Infrastructure project.

Thank you so much for your help! It works now!

**Comment by fiseni** [comment]
I'm glad you resolved it. Closing the issue.
